libcontainer/seccomp/patchbpf/enosys_linux.go

```diff
@@ -522,10 +522,30 @@ func generateEnosysStub(lastSyscalls lastSyscallMap) ([]bpf.Instruction, error)
 
 	// Prepend the load instruction for the architecture.
 	programTail = append([]bpf.Instruction{
-		// load [4]
+		// load [4] (architecture)
 		bpf.LoadAbsolute{Off: 4, Size: 4}, // NOTE: We assume sizeof(int) == 4.
 	}, programTail...)
 
+	// Return -ENOSYS for setup(2) -- on s390x (and possibly other
+	// platforms) this syscall is used for multiplexing "large syscall
+	// number" syscalls, but if the syscall number is not known to the
+	// kernel then the syscall number is left unchanged (and is thus
+	// sys_num=0). The actual setup(2) syscall is never used by userspace
+	// anymore (and hasn't existed for decades) outside of this multiplexing
+	// scheme so returning -ENOSYS is fine.
+	programTail = append([]bpf.Instruction{
+		// load [0] (syscall number)
+		bpf.LoadAbsolute{Off: 0, Size: 4}, // NOTE: We assume sizeof(int) == 4.
+		// jne 0,1
+		bpf.JumpIf{
+			Cond:     bpf.JumpNotEqual,
+			Val:      0,
+			SkipTrue: 1,
+		},
+		// ret [ENOSYS]
+		bpf.RetConstant{Val: retErrnoEnosys},
+	}, programTail...)
+
```

Comment:
- This should be appended only for `GOARCH == s390x` ?
- Could you define a `const` for the syscall number of `setup(2)`?

Reply:
1. That would be one way but I suspect we need the same hack for s390 as well (I'm not sure -- I don't have an s390 box).

I'm currently working on making this addition architecture-independent so that we add this rule for every architecture (since `setup(2)` isn't used by any user code intentionally we should be able to always return `-ENOSYS` even for architectures that don't do this weird multiplexing thing).

Reply:
libseccomp doesn't define setup for any architecture so I just hard-coded it (with a defined const) for s390 and s390x. The only other architecture which has a setup syscall is MIPS (and only on Linux-o32). So this should cover all of our bases.